test-runtime/build.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of subxt.
+//
+// subxt is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// subxt is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with subxt.  If not, see <http://www.gnu.org/licenses/>.
+
+use serde_json::json;
+use std::{
+    env,
+    fs,
+    net::TcpListener,
+    path::Path,
+    process::Command,
+    sync::atomic::{
+        AtomicU16,
+        Ordering,
+    },
+    thread,
+    time,
+};
+
+static SUBSTRATE_BIN_ENV_VAR: &str = "SUBSTRATE_NODE_PATH";
+
+fn main() {
+    // Select substrate binary to run based on env var.
+    let substrate_bin =
+        env::var(SUBSTRATE_BIN_ENV_VAR).unwrap_or_else(|_| "substrate".to_owned());
+
+    // Run binary.
+    let port = next_open_port()
+        .expect("Cannot spawn substrate: no available ports in the given port range");
+    let cmd = Command::new(&substrate_bin)
+        .arg("--dev")
+        .arg("--tmp")
+        .arg(format!("--rpc-port={}", port))
+        .spawn();
+    let mut cmd = match cmd {
+        Ok(cmd) => cmd,
+        Err(e) => {
+            panic!("Cannot spawn substrate command '{}': {}", substrate_bin, e)
+        }
+    };
+
+    // Download metadata from binary; retry until successful, or a limit is hit.
+    let res: serde_json::Value = {
+        const MAX_RETRIES: usize = 20;
+        let mut retries = 0;
+        let mut wait_secs = 1;
+        loop {
+            if retries >= MAX_RETRIES {
+                panic!("Cannot connect to substrate node after {} retries", retries);
+            }
+            let res = reqwest::blocking::Client::new()
+                .post(format!("http://localhost:{}", port))
+                .json(&json!({
+                   "id": 1,
+                   "jsonrpc": "2.0",
+                   "method": "state_getMetadata",
+                }))
+                .send();
```

Comment:
If we use `subxt::RpcClient` as mentioned above, you could just do e.g. https://github.com/paritytech/subxt/blob/7e26f9559c26a9644eb4cd189e06ab1ad8bc16dd/src/rpc.rs#L413